prometheus/pkg/reconciler/prometheussource.go

```diff
@@ -0,0 +1,399 @@
+// converge the two. It then updates the Status block of the PrometheusSource
+// resource with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		r.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the PrometheusSource resource with this namespace/name
+	original, err := r.prometheussourceLister.PrometheusSources(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("PrometheusSource key in work queue no longer exists", zap.Any("key", key))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	prometheussource := original.DeepCopy()
+
+	// Reconcile this copy of the PrometheusSource and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(ctx, prometheussource)
+	if err != nil {
+		logging.FromContext(ctx).Warn("Error reconciling PrometheusSource", zap.Error(err))
+	} else {
+		logging.FromContext(ctx).Debug("PrometheusSource reconciled")
+		r.Recorder.Eventf(prometheussource, corev1.EventTypeNormal, prometheussourceReconciled, `PrometheusDbSource reconciled: "%s/%s"`, prometheussource.Namespace, prometheussource.Name)
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, prometheussource.DeepCopy()); updateStatusErr != nil {
+		logging.FromContext(ctx).Warn("Failed to update the PrometheusSource", zap.Error(err))
+		r.Recorder.Eventf(prometheussource, corev1.EventTypeWarning, prometheussourceUpdateStatusFailed, "Failed to update PrometheusSource's status: %v", err)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready:
+	return err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PrometheusSource) error {
+	source.Status.InitializeConditions()
+
+	sinkObjRef := source.Spec.Sink
+	if sinkObjRef.Namespace == "" {
+		sinkObjRef.Namespace = source.Namespace
+	}
+
+	sourceDesc := source.Namespace + "/" + source.Name + ", " + source.GroupVersionKind().String()
+	sinkURI, err := r.sinkReconciler.GetSinkURI(sinkObjRef, source, sourceDesc)
+	if err != nil {
+		source.Status.MarkNoSink("NotFound", "")
+		return err
+	}
+	source.Status.MarkSink(sinkURI)
+
+	ra, err := r.createReceiveAdapter(ctx, source, sinkURI)
+	if err != nil {
+		r.Logger.Error("Unable to create the receive adapter", zap.Error(err))
+		return err
+	}
+	// Update source status// Update source status
+	source.Status.PropagateDeploymentAvailability(ra)
+
+	err = r.reconcileEventTypes(ctx, source)
+	if err != nil {
+		source.Status.MarkNoEventTypes("EventTypesReconcileFailed", "")
+		return err
+	}
+	source.Status.MarkEventTypes()
+
+	return nil
+}
+
+func (r *Reconciler) getReceiveAdapterImage() string {
```

Comment:
Heads-up: when this [PR](https://github.com/knative/eventing/pull/2030) gets merged we should do the same, which is to set the adapter image in `controller`

Reply:
Sure thing.

Reply:
@syedriko @lionelvillard it got merged!